### PR TITLE
build: enable gn check for //electron:electron_lib

### PR DESCRIPTION
**Pull request by deepak1556**
#### Description of Change

- [x] Fix gn check errors for `//electron`
- [x] Add --gn-check lint option
- [x] Add check to ci

The headers included from some private targets have been suffixed with `nogncheck`, these targets are known to be built as a transitive dep, so lets avoid a visibility patch in chromium. Also this gives a good idea of private headers we depend on, lets try to minimize them. 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: no-notes

### CODE REVIEW

File: atom/browser/api/atom_api_app.cc [hunk 1]
```diff
@@ -34,8 +34,8 @@
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/icon_manager.h"
 #include "chrome/common/chrome_paths.h"
-#include "content/browser/gpu/compositor_util.h"
-#include "content/browser/gpu/gpu_data_manager_impl.h"
+#include "content/browser/gpu/compositor_util.h"        // nogncheck
+#include "content/browser/gpu/gpu_data_manager_impl.h"  // nogncheck
```
Comment (nornagon) [inline atom/browser/api/atom_api_app.cc]: can we have additional comments for nogncheck'd files re: why we can't just have the correct dependencies?

Comment (nornagon) [reply atom/browser/api/atom_api_app.cc]: er nvm, you commented on this in the PR description.

**Comment by deepak1556** [comment]
@nornagon can you review this again for precommit hook changes.

File: script/lint.js [hunk 2]
```diff
@@ -200,17 +235,21 @@ async function main () {
...
 
   const linters = LINTERS.filter(x => opts[x.key])
 
   for (const linter of linters) {
-    const filenames = await findFiles(opts, linter)
-    if (filenames.length) {
-      if (opts.verbose) { console.log(`linting ${filenames.length} ${linter.key} ${filenames.length === 1 ? 'file' : 'files'}`) }
-      linter.run(opts, filenames)
+    if (linter.key === 'gn-check') {
+      linter.run(opts)
```
Comment (nornagon) [inline script/lint.js]: we should only run `gn check` if a *.{gn,gni,h,c,cc,m,mm} file changes.

File: package.json [hunk 3]
```diff
@@ -102,7 +102,8 @@
     ],
     "*.{gn,gni}": [
       "python script/run-gn-format.py",
-      "git add"
+      "git add",
+      "node script/lint.js --gn-check --"
```
[the remark this answers is not in the document]
Comment (deepak1556) [reply package.json]: > can `gn check` change a file?

Nope it doesn't, the errors needs to be fixed manually. Thats why I added it after `git add`

File: script/lint.js (hunk 2 above)
Comment (deepak1556) [reply script/lint.js]: Adding this to `lint-staged` hook should take care of it.

**Comment by deepak1556** [comment]
gn gen takes considerable time that might be intrusive for a pre-commit hook even if its run on a cached dir. Moreover we need to test both testing and debug config, maybe we should just have this as a pre-push hook instead ?

**Comment by nornagon** [comment]
Could we run it on a bot?

**Comment by deepak1556** [comment]
Yeah that would make things a lot easier, if I add them as a step in the testing and build CI runs, since this needs a complete source checkout and should be tested across all platforms. /cc @jkleinsc thoughts on this ?

**Comment by MarshallOfSound** [comment]
@deepak1556 How long would this take to run, and would the output / result change on different platforms?

**Comment by jkleinsc** [comment]
@deepak1556 is there any reason to run it against multiple architectures or could we just run it on one?

**Comment by deepak1556** [comment]
> How long would this take to run

```
Header dependency check OK

real	0m10.266s
user	0m25.371s
sys	0m5.329s
```

> would the output / result change on different platforms?

Yup only when you have platform specific includes and deps.

**Comment by MarshallOfSound** [comment]
Hm yeah that's probably a bit too long for pre-commit 🤔 I think letting the `lint` stage catch this on CircleCI would be enough

**Comment by deepak1556** [comment]
> I think letting the `lint` stage catch this on CircleCI would be enough

I don't think the lint ci has the complete src/ checkout, this check needs it. Any concern in adding this as a step in the build CI runs ? Just need the following variations, the cpu arch doesn't matter since we don't include deps based on that condition.

```
win-testing
win-debug
linux-testing
linux-debug
mac-testing
mac-debug
```

**Comment by MarshallOfSound** [comment]
@deepak1556 I'm 👍 on adding it to those jobs you mentioned but **only** if it is the first step in those jobs.  i.e. the check runs before we try to build Electron.  Otherwise we might have to wait an hour just to find out about a lint failure 👍

**Comment by nornagon** [comment]
Perhaps on circleci it would make sense to run it in parallel with the build? that way if lint fails, you still get to see the result of the tests.

**Comment by deepak1556** [comment]
I could use some help with fixing gn check errors on linux, thanks! will push fix for errors on windows shortly.

File: script/lint.js [hunk 4]
```diff
@@ -114,12 +114,47 @@ const LINTERS = [ {
       process.exit(1)
     }
   }
+}, {
+  key: 'gn-check',
+  run: (opts) => {
+    const env = Object.assign({
```
[the remark this answers is not in the document]
Comment (nornagon) [reply script/lint.js]: still think spread > Object.assign :)

File: script/lint.js [hunk 5]
```diff
@@ -199,18 +239,23 @@ async function findFiles (args, linter) {
 async function main () {
   const opts = parseCommandLine()
 
-  // no mode specified? run 'em all
-  if (!opts['c++'] && !opts.javascript && !opts.python && !opts.gn) {
-    opts['c++'] = opts.javascript = opts.python = opts.gn = true
+  // no mode specified? run 'em all excpet gn-check,
```
Comment (nornagon) [inline script/lint.js]:
```suggestion
  // no mode specified? run 'em all except gn-check,
```

File: script/lint.js (hunk 2 above)
Comment (nornagon) [reply script/lint.js]: since this is running in CI now, this is moot.

File: script/lint.js [hunk 6]
```diff
@@ -41,6 +41,32 @@ function spawnAndCheckExitCode (cmd, args, opts) {
   if (status) process.exit(status)
 }
 
+function gnCheckConfig (config, env) {
```
Comment (nornagon) [inline script/lint.js]: i'm inclined to drop this from the script now that it's just a simple `gn check` in CI.

Comment (deepak1556) [reply script/lint.js]: Yup I would be in favor of that, wasn't sure if users wanted to a simple one liner to replicate the failure they saw in ci.

Comment (nornagon) [reply script/lint.js]: i think this one-liner will probably cause more problems than it solves. `gn check out/Debug //electron:electron_lib` is already a one-liner :)

**Comment by deepak1556** [comment]
@jkleinsc @MarshallOfSound can you review the ci config changes. Thanks!

**Comment by jkleinsc** [comment]
@nornagon since you suggested running gn check in parallel, I'm wondering if you still stand by that given how quickly gn check runs?  We are incurring an extra 5 minutes of compute time to run a 15 second check.

**Comment by nornagon** [comment]
the reason i suggested it being a parallel build is so that if the `gn check` fails you also get build/test results. it sucks when lint prevents your build from happening, especially when it takes 15+ minutes for that build to even start.

my desired solutions, from most to least desirable:
1. have the `check` step run on the same checkout as the build/tests, but have it show up as a separate check from the build/test results in the GitHub UI
2. have the `check` step run on the same checkout as the build/tests, but let the build keep going and only report failure at the end
3. have the `check` step block the build

if we end up having to go with (3) it wouldn't be the worst thing, just a bit of a bummer.

**Comment by jkleinsc** [comment]
I did not realize we were trying to build even if gn check failed.  If that is the case, I'm fine with the PR as is.

**Pull request merged**